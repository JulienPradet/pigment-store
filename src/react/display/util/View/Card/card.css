.container {
    border: 1px solid rgba(200, 200, 200, 0.5);
    margin: 2em 0;
}

.title {
    padding: 1em;
    border-bottom: 1px solid rgba(200, 200, 200, 0.5);
}

.title h2 {
    margin: 0;
}

.content {
    padding: 1em;
}
