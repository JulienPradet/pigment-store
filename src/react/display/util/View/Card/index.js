import React from 'react'
import Collapse from 'react-collapse'
import {compose, branch, withHandlers, withState} from 'recompose'
import styles from './card.css'

const Title = ({closed, onClick, children}) => <div className={styles.title}>
  <div className={styles.titleContent}>
    {children}
  </div>
  {onClick
    ? <div className={styles.titleButton}>
      <button onClick={onClick}>
        <i className='material-icons'>{closed ? 'keyboard_arrow_down' : 'keyboard_arrow_up'}</i>
      </button>
    </div>
    : null}
</div>

const Content = ({children}) => <div className={styles.content}>{children}</div>

const Card = ({id, closed, title, subTitle, children, toggle}) => {
  if (children && typeof children === 'object' && children.length) {
    children = children.filter((child) => child).length > 0 ? children : null
  }

  return <div className={styles.container} id={id}>
    {title || subTitle
      ? <div className={styles.titleContainer} onClick={toggle}>
        <Title closed={closed} onClick={toggle}>{title}</Title>
        {subTitle ? <Content>{subTitle}</Content> : null}
      </div>
      : null}
    {children
      ? <Collapse
        isOpened={!closed}
      >
        <div className={styles.collapsibleContent}>
          {children}
        </div>
      </Collapse>
      : null}
  </div>
}

const SmartCard = compose(
  branch(
    (props) => props.closable,
    (BaseComponent) => BaseComponent,
    () => Card
  ),
  withState('closed', 'setClosed', (props) => props.defaultClosed),
  withHandlers({
    toggle: (props) => () => props.setClosed(!props.closed)
  })
)(Card)

SmartCard.displayName = 'Card'

export default SmartCard
