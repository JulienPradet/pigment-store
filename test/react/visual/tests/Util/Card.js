import React from 'react'
import PigmentStore from '../../../../../dist/index'
import Card from '../../../../../src/react/display/util/View/Card/index'

const cardContent = {
  title: 'Card Title',
  subTitle: 'Short summary of the card',
  children: 'Actual description of the card. The content can be very versatile and as long as needed.'
}

const Span = (props) => <span {...props} />

export default PigmentStore.React.describe('Card', Card)
  .setDescription(`
A card is a block containing a title and a content that serves as a detailed description.

The main feature of a Card is that the content can be collapsed.
  `)
  .feature(
    'Default',
    cardContent
  )
  .feature(
    'Closable card',
    Object.assign({}, cardContent, {closable: true})
  )
  .feature(
    'Closable card initialized as closed',
    Object.assign({}, cardContent, {closable: true, defaultClosed: true})
  )
  .feature(
    'Card with a component as a title',
    Object.assign({}, cardContent, {
      closable: true,
      defaultClosed: true,
      title: <Span style={{color: 'red'}}>Title</Span>
    })
  )
