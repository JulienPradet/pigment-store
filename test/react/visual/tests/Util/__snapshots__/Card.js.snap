exports[`Card Card with a component as a title 1`] = `
<div
  className="container"
  id={undefined}>
  <div
    className="titleContainer"
    onClick={[Function]}>
    <div
      className="title">
      <div
        className="titleContent">
        <span
          style={
            Object {
              "color": "red",
            }
          }>
          Title
        </span>
      </div>
      <div
        className="titleButton">
        <button
          onClick={[Function]}>
          <i
            className="material-icons">
            keyboard_arrow_down
          </i>
        </button>
      </div>
    </div>
    <div
      className="content">
      Short summary of the card
    </div>
  </div>
  <div />
</div>
`;

exports[`Card Closable card 1`] = `
<div
  className="container"
  id={undefined}>
  <div
    className="titleContainer"
    onClick={[Function]}>
    <div
      className="title">
      <div
        className="titleContent">
        Card Title
      </div>
      <div
        className="titleButton">
        <button
          onClick={[Function]}>
          <i
            className="material-icons">
            keyboard_arrow_up
          </i>
        </button>
      </div>
    </div>
    <div
      className="content">
      Short summary of the card
    </div>
  </div>
  <div>
    <div
      className="collapsibleContent">
      Actual description of the card. The content can be very versatile and as long as needed.
    </div>
  </div>
</div>
`;

exports[`Card Closable card initialized as closed 1`] = `
<div
  className="container"
  id={undefined}>
  <div
    className="titleContainer"
    onClick={[Function]}>
    <div
      className="title">
      <div
        className="titleContent">
        Card Title
      </div>
      <div
        className="titleButton">
        <button
          onClick={[Function]}>
          <i
            className="material-icons">
            keyboard_arrow_down
          </i>
        </button>
      </div>
    </div>
    <div
      className="content">
      Short summary of the card
    </div>
  </div>
  <div />
</div>
`;

exports[`Card Default 1`] = `
<div
  className="container"
  id={undefined}>
  <div
    className="titleContainer"
    onClick={undefined}>
    <div
      className="title">
      <div
        className="titleContent">
        Card Title
      </div>
    </div>
    <div
      className="content">
      Short summary of the card
    </div>
  </div>
  <div>
    <div
      className="collapsibleContent">
      Actual description of the card. The content can be very versatile and as long as needed.
    </div>
  </div>
</div>
`;
